File: backend/app/schemas/meme.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict


class MemeCharacterRead(BaseModel):
    id: int
    slug: str
    name: str
    avatar_url: str | None = None
    theme_color: str | None = None

    model_config = ConfigDict(from_attributes=True)


class MemeRead(BaseModel):
    id: int
    slug: str
    title: str
    description: str | None = None
    image_url: str
    file_type: str
    character_id: int | None = None
    character_ids: list[int] = []
    source_name: str | None = None
    source_url: str | None = None
    author_name: str | None = None
    author_id: int | None = None
    author_uid: str | None = None
    view_count: int
    download_count: int
    sort_order: int
    is_featured: bool
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class MemeDetailRead(MemeRead):
    character: MemeCharacterRead | None = None

File: frontend/src/types/meme.ts
export interface Meme {
  id: number
  slug: string
  title: string
  description: string | null
  image_url: string
  file_type: string
  character_id: number | null
  character_ids: number[]
  source_name: string | null
  source_url: string | null
  author_name: string | null
  author_id: number | null
  author_uid: string | null
  view_count: number
  download_count: number
  sort_order: number
  is_featured: boolean
  created_at: string
  updated_at: string
}

export interface MemeCharacter {
  id: number
  slug: string
  name: string
  avatar_url: string | null
  theme_color: string | null
}

export interface MemeDetail extends Meme {
  character: MemeCharacter | null
}
